File: scripts/crossword-generator.py
import json
import random
import argparse
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Optional
import requests
from bs4 import BeautifulSoup


# ============================================================
# KONFIGURACJA
# ============================================================

DEFAULT_INPUT = "slownik_bez_rzadkich.json"
DEFAULT_OUTPUT = "crossword.json"

SIZE = 5

MIN_WORD_LENGTH = 3
MAX_WORD_LENGTH = 5

# Ile prób stworzenia planszy
MAX_ATTEMPTS = 500

# Ile różnych układów czarnych pól próbujemy
MAX_BLOCK_PATTERNS = 100


# ============================================================
# SLOT
# ============================================================

@dataclass
class Slot:
    row: int
    col: int
    length: int
    direction: str

    @property
    def cells(self):
        result = []

        if self.direction == "horizontal":
            for i in range(self.length):
                result.append(
                    (self.row, self.col + i)
                )
        else:
            for i in range(self.length):
                result.append(
                    (self.row + i, self.col)
                )

        return result


# ============================================================
# WCZYTYWANIE SŁÓW
# ============================================================
def _clean_clue(text, word=None):
    """Czyści definicję i usuwa techniczne elementy ze źródła."""
    if not text:
        return None

    text = BeautifulSoup(text, "html.parser").get_text(" ", strip=True)
    text = re.sub(r"\s+", " ", text).strip()

    # Nie chcemy przypadkowo zapisać elementów nawigacji strony.
    bad_exact = {
        "komentarze",
        "brak",
        "brak definicji",
        "więcej",
        "lista",
        "słownik języka polskiego sjp",
        "info",
    }
    if text.lower() in bad_exact:
        return None

    # Jeśli tekst zawiera numery (np "1. definicja 2. inna"), bierz tylko pierwszą
    match = re.match(r"^(\d+\.\s*)?(.+?)(?:\s*\d+\.|$)", text)
    if match:
        text = match.group(2).strip()

    # Usuń numerację typu "1. " lub "(1)" z początku i końca
    text = re.sub(r"^\d+\.\s*", "", text).strip()
    text = re.sub(r"^\(\d+(?:\.\d+)?\)\s*", "", text).strip()
    text = re.sub(r"\s*\(\d+(?:\.\d+)?\)$", "", text).strip()

    if not text:
        return None

    # Odrzuć clue'y, które wyglądają na junk - same słowa techniczne z numeracją
    if re.match(r"^(info|undefined|brak|puste)\s*(?:\(\d+\))?$", text.lower()):
        return None

    # Jeśli mamy słowo klucza, sprawdź czy definicja go nie zawiera
    if word:
        word_lower = word.lower()
        text_lower = text.lower()
        
        # Unikaj definicji, które zaczynają się od samego słowa lub jego formy
        if text_lower.startswith(word_lower) or text_lower.startswith(f"rodzaj {word_lower}") or \
           text_lower.startswith(f"forma {word_lower}"):
            return None
        
        # Unikaj definicji, które są tylko powtórzeniem słowa
        if text_lower == word_lower:
            return None

    if len(text) > 120:
        text = text[:117].rsplit(" ", 1)[0] + "..."

    return text


def _get_clue_sjp(word):
    """
    Pobiera definicję z SJP.pl.

    Szuka konkretnie w paragrafach i listach, unikając
    technicznych elementów strony.
    """
    url = f"https://sjp.pl/{word}"

    try:
        response = requests.get(
            url,
            timeout=10,
            headers={
                "User-Agent": (
                    "Mozilla/5.0 "
                    "(Windows NT 10.0; Win64; x64) "
                    "AppleWebKit/537.36 "
                    "Chrome/140.0 Safari/537.36"
                )
            },
        )

        if response.status_code != 200:
            return None

        soup = BeautifulSoup(response.text, "html.parser")

        # Szukamy sekcji "znaczenie:" w tekście
        marker = None
        for element in soup.find_all(string=re.compile(r"^\s*znaczenie\s*:", re.I)):
            marker = element.parent
            break

        if marker is None:
            return None

        # Szukamy tylko w paragrafach i elementach listy (dd, li)
        # Po znalezieniu markera "znaczenie:"
        for element in marker.find_all_next():
            # Jeśli natrafimy na nagłówek, kończymy szukanie
            if element.name in {"h1", "h2", "h3"}:
                break

            # Bierzemy tekst tylko z rzeczywistych elementów zawartości
            if element.name not in {"p", "dd", "li", "div"}:
                continue

            value = element.get_text(" ", strip=True)
            if not value:
                continue

            # Usuń marker "znaczenie:" jeśli nadal tam jest
            value = re.sub(r"^\s*znaczenie\s*:\s*", "", value, flags=re.I).strip()
            
            # Oczyść i zwaliduj - przekaż słowo, żeby sprawdzić czy się nie powtarza
            value = _clean_clue(value, word=word)

            if value and len(value) >= 5:
                return value

        return None

    except Exception:
        return None


def _get_clue_wiktionary(word):
    """
    Pobiera pierwszą polską definicję z Wikisłownika.
    Celujemy bezpośrednio w sekcję "znaczenia", dzięki czemu
    nie łapiemy elementów menu strony.
    """
    url = f"https://pl.wiktionary.org/wiki/{word}"

    try:
        response = requests.get(
            url,
            timeout=10,
            headers={
                "User-Agent": (
                    "Mozilla/5.0 "
                    "(Windows NT 10.0; Win64; x64) "
                    "AppleWebKit/537.36 "
                    "Chrome/140.0 Safari/537.36"
                )
            },
        )

        if response.status_code != 200:
            return None

        soup = BeautifulSoup(response.text, "html.parser")

        # Najpierw znajdź sekcję języka polskiego.
        polish_section = None

        for heading in soup.find_all(["h2", "h3"]):
            heading_text = heading.get_text(" ", strip=True).lower()

            if "język polski" in heading_text:
                polish_section = heading
                break

        if polish_section is None:
            return None

        # W obrębie polskiej sekcji znajdź nagłówek "znaczenia".
        meanings_heading = None

        for element in polish_section.find_all_next(["h3", "h4"], limit=30):
            text_value = element.get_text(" ", strip=True).lower()

            if "znaczenia" == text_value:
                meanings_heading = element
                break

            # Jeżeli trafiliśmy na kolejną dużą sekcję, kończymy.
            if element.name == "h2":
                break

        if meanings_heading is None:
            return None

        # Pierwsze <li> po "znaczenia" zawiera zwykle definicję.
        for element in meanings_heading.find_all_next():
            if element.name in {"h2", "h3"}:
                break

            if element.name == "li":
                clue = _clean_clue(element.get_text(" ", strip=True), word=word)

                if clue:
                    return clue

        return None

    except Exception:
        return None


def get_clue(word):
    """
    Pobiera prawdziwą polską definicję słowa.

    Kolejność:
      1. lokalny cache
      2. SJP.pl
      3. Wikisłownik

    Cache zapisujemy w clue_cache.json, żeby przy kolejnych
    uruchomieniach nie odpytwać ponownie tych samych stron.
    """
    cache_file = Path("clue_cache.json")

    try:
        if cache_file.exists():
            with cache_file.open("r", encoding="utf-8") as f:
                cache = json.load(f)
        else:
            cache = {}
    except Exception:
        cache = {}

    word = word.strip().lower()

    if word in cache:
        cached = cache[word]
        return cached if cached else None

    clue = _get_clue_sjp(word)

    if clue is None:
        clue = _get_clue_wiktionary(word)

    # Zapamiętaj również brak definicji, żeby nie odpytwać
    # ponownie tego samego słowa przy kolejnym uruchomieniu.
    cache[word] = clue

    try:
        with cache_file.open("w", encoding="utf-8") as f:
            json.dump(
                cache,
                f,
                ensure_ascii=False,
                indent=2,
            )
    except Exception:
        pass

    return clue


def load_words(filename):
    with open(filename, "r", encoding="utf-8") as f:
        data = json.load(f)

    words = set()

    for letter_data in data.values():

        for length in ("3", "4", "5"):

            for word in letter_data.get(length, []):

                word = word.strip().lower()

                if not word:
                    continue

                if len(word) < MIN_WORD_LENGTH:
                    continue

                if len(word) > MAX_WORD_LENGTH:
                    continue

                words.add(word)

    print(f"Wczytano {len(words):,} unikalnych słów.")

    return sorted(words)


# ============================================================
# WZORZEC PLANSZY
# ============================================================

def generate_block_pattern(rng):
    """
    Generuje planszę z czarnymi polami.

    Używamy symetrii rotacyjnej:
    jeśli (r,c) jest czarne, to (4-r,4-c)
    również jest czarne.

    Dzięki temu plansza wygląda bardziej
    jak klasyczna krzyżówka.
    """

    grid = [
        [False for _ in range(SIZE)]
        for _ in range(SIZE)
    ]

    # 5x5 - losujemy kilka pól
    candidates = []

    for r in range(SIZE):
        for c in range(SIZE):

            rr = SIZE - 1 - r
            cc = SIZE - 1 - c

            # środek zostawiamy wolny
            if (r, c) == (2, 2):
                continue

            if (r, c) > (rr, cc):
                continue

            candidates.append((r, c))

    rng.shuffle(candidates)

    # Raczej niewiele czarnych pól.
    block_count = rng.randint(0, 6)

    used = 0

    for r, c in candidates:

        rr = SIZE - 1 - r
        cc = SIZE - 1 - c

        if grid[r][c]:
            continue

        if grid[rr][cc]:
            continue

        grid[r][c] = True
        grid[rr][cc] = True

        used += 1

        if used >= block_count:
            break

    return grid


# ============================================================
# SPRAWDZENIE WZORCA
# ============================================================

def extract_slots(blocks):
    """
    Znajduje wszystkie poziome i pionowe sloty.

    Slot musi mieć minimum 3 pola.
    """

    slots = []

    # --------------------------------------------------------
    # POZIOMO
    # --------------------------------------------------------

    for r in range(SIZE):

        c = 0

        while c < SIZE:

            if blocks[r][c]:
                c += 1
                continue

            start = c

            while c < SIZE and not blocks[r][c]:
                c += 1

            length = c - start

            if length >= MIN_WORD_LENGTH:
                slots.append(
                    Slot(
                        row=r,
                        col=start,
                        length=length,
                        direction="horizontal"
                    )
                )

    # --------------------------------------------------------
    # PIONOWO
    # --------------------------------------------------------

    for c in range(SIZE):

        r = 0

        while r < SIZE:

            if blocks[r][c]:
                r += 1
                continue

            start = r

            while r < SIZE and not blocks[r][c]:
                r += 1

            length = r - start

            if length >= MIN_WORD_LENGTH:
                slots.append(
                    Slot(
                        row=start,
                        col=c,
                        length=length,
                        direction="vertical"
                    )
                )

    return slots


def valid_block_pattern(blocks):
    """
    Odrzuca układy, które mają wolne pola tworzące
    jedno- lub dwuliterowe fragmenty.

    Każdy slot musi mieć długość 3-5.
    """

    slots = extract_slots(blocks)

    # Wszystkie niezamknięte fragmenty muszą być slotami.
    # Jeśli gdzieś powstaje długość 1/2 -> odrzucamy.
    for r in range(SIZE):

        c = 0

        while c < SIZE:

            if blocks[r][c]:
                c += 1
                continue

            start = c

            while c < SIZE and not blocks[r][c]:
                c += 1

            length = c - start

            if length not in (0, 3, 4, 5):
                return False

    for c in range(SIZE):

        r = 0

        while r < SIZE:

            if blocks[r][c]:
                r += 1
                continue

            start = r

            while r < SIZE and not blocks[r][c]:
                r += 1

            length = r - start

            if length not in (0, 3, 4, 5):
                return False

    return True


# ============================================================
# KANDYDACI
# ============================================================

def get_candidates(slot, grid, words_by_length):
    """
    Zwraca słowa pasujące do aktualnego slotu.
    """

    candidates = words_by_length.get(slot.length, [])

    result = []

    for word in candidates:

        fits = True

        for index, (r, c) in enumerate(slot.cells):

            existing = grid[r][c]

            if existing is not None:
                if existing != word[index]:
                    fits = False
                    break

        if fits:
            result.append(word)

    return result


# ============================================================
# WSTAWIANIE SŁOWA
# ============================================================

def place_word(slot, word, grid):
    """
    Wstawia słowo i zwraca informację,
    które pola zmieniliśmy.
    """

    changed = []

    for index, (r, c) in enumerate(slot.cells):

        if grid[r][c] is None:
            grid[r][c] = word[index]
            changed.append((r, c))

    return changed


def undo(changed, grid):
    for r, c in changed:
        grid[r][c] = None


# ============================================================
# BACKTRACKING
# ============================================================

def solve_slots(slots, grid, words_by_length, used_words):
    """
    Backtracking.

    Za każdym razem wybieramy slot z najmniejszą
    liczbą możliwych słów.

    To bardzo mocno ogranicza liczbę prób.
    """

    if not slots:
        return True

    # --------------------------------------------------------
    # Wybierz najbardziej ograniczony slot
    # --------------------------------------------------------

    best_slot = None
    best_candidates = None

    for slot in slots:

        candidates = get_candidates(
            slot,
            grid,
            words_by_length
        )

        # Nie używamy tego samego słowa ponownie.
        candidates = [
            word
            for word in candidates
            if word not in used_words
        ]

        if not candidates:
            return False

        if (
            best_candidates is None
            or len(candidates) < len(best_candidates)
        ):
            best_slot = slot
            best_candidates = candidates

    # Losowa kolejność -> różne krzyżówki
    random.shuffle(best_candidates)

    remaining_slots = [
        slot
        for slot in slots
        if slot is not best_slot
    ]

    # --------------------------------------------------------
    # Próby
    # --------------------------------------------------------

    for word in best_candidates:

        changed = place_word(
            best_slot,
            word,
            grid
        )

        used_words.add(word)

        if solve_slots(
            remaining_slots,
            grid,
            words_by_length,
            used_words
        ):
            return True

        used_words.remove(word)

        undo(
            changed,
            grid
        )

    return False


# ============================================================
# WALIDACJA ROZWIĄZANIA
# ============================================================

def validate_solution(slots, grid, used_words):
    """
    Dodatkowa kontrola.
    """

    if len(used_words) != len(slots):
        return False

    for slot in slots:

        chars = []

        for r, c in slot.cells:

            if grid[r][c] is None:
                return False

            chars.append(
                grid[r][c]
            )

        word = "".join(chars)

        if word not in used_words:
            return False

    return True


# ============================================================
# FORMATOWANIE
# ============================================================

def create_output(blocks, slots, grid):

    words = []

    for slot in slots:

        word = "".join(
            grid[r][c]
            for r, c in slot.cells
        )
        clue = get_clue(word)
        entry = {
            "word": word,
            "row": slot.row,
            "col": slot.col,
            "length": slot.length,
            "direction": slot.direction,
            "clue": clue if clue else "Brak definicji"
        }

        words.append(entry)

    return {
        "size": SIZE,

        "grid": [
            [
                "#" if blocks[r][c]
                else grid[r][c]
                for c in range(SIZE)
            ]
            for r in range(SIZE)
        ],

        "words": words
    }


# ============================================================
# WYŚWIETLANIE
# ============================================================

def print_board(data):

    print()
    print("PLANSZA")
    print("-" * 20)

    for row in data["grid"]:

        print(
            " ".join(
                cell if cell != "#" else "■"
                for cell in row
            )
        )

    print()

    print("POZIOMO:")

    for word in data["words"]:
        if word["direction"] == "horizontal":
            print(
                f'  {word["word"]} '
                f'({word["row"] + 1},{word["col"] + 1})'
            )

    print()

    print("PIONOWO:")

    for word in data["words"]:
        if word["direction"] == "vertical":
            print(
                f'  {word["word"]} '
                f'({word["row"] + 1},{word["col"] + 1})'
            )


# ============================================================
# GENEROWANIE
# ============================================================

def generate_crossword(words, seed=None):

    rng = random.Random(seed)

    words_by_length = {
        3: [],
        4: [],
        5: []
    }

    for word in words:
        words_by_length[len(word)].append(word)

    # --------------------------------------------------------
    # Próby
    # --------------------------------------------------------

    for attempt in range(1, MAX_ATTEMPTS + 1):

        blocks = generate_block_pattern(rng)

        if not valid_block_pattern(blocks):
            continue

        slots = extract_slots(blocks)

        # Musimy mieć zarówno poziome,
        # jak i pionowe hasła.
        horizontal_count = sum(
            1
            for slot in slots
            if slot.direction == "horizontal"
        )

        vertical_count = sum(
            1
            for slot in slots
            if slot.direction == "vertical"
        )

        if horizontal_count == 0:
            continue

        if vertical_count == 0:
            continue

        # ----------------------------------------------------
        # Pusta plansza
        # ----------------------------------------------------

        grid = [
            [
                None if not blocks[r][c]
                else "#"
                for c in range(SIZE)
            ]
            for r in range(SIZE)
        ]

        used_words = set()

        # ----------------------------------------------------
        # Rozwiązanie
        # ----------------------------------------------------

        if solve_slots(
            slots,
            grid,
            words_by_length,
            used_words
        ):

            if not validate_solution(
                slots,
                grid,
                used_words
            ):
                continue

            return create_output(
                blocks,
                slots,
                grid
            )

    return None


# ============================================================
# MAIN
# ============================================================

def main():

    parser = argparse.ArgumentParser(
        description="Generator polskiej krzyżówki 5x5"
    )

    parser.add_argument(
        "--input",
        default=DEFAULT_INPUT,
        help="Plik JSON ze słowami"
    )

    parser.add_argument(
        "--output",
        default=DEFAULT_OUTPUT,
        help="Plik wynikowy JSON"
    )

    parser.add_argument(
        "--seed",
        type=int,
        default=None,
        help="Seed losowania"
    )

    args = parser.parse_args()

    # --------------------------------------------------------
    # Słownik
    # --------------------------------------------------------

    words = load_words(
        args.input
    )

    # --------------------------------------------------------
    # Generowanie
    # --------------------------------------------------------

    print("Generowanie krzyżówki...")

    crossword = generate_crossword(
        words,
        seed=args.seed
    )

    if crossword is None:

        print()
        print(
            "[ERROR] Nie udało się wygenerować krzyżówki."
        )

        print(
            "Spróbuj innego --seed albo zwiększ "
            "MAX_ATTEMPTS."
        )

        return

    # --------------------------------------------------------
    # Zapis
    # --------------------------------------------------------

    with open(
        args.output,
        "w",
        encoding="utf-8"
    ) as f:

        json.dump(
            crossword,
            f,
            ensure_ascii=False,
            indent=2
        )

    print_board(crossword)

    print()
    print(
        f"Zapisano do: {args.output}"
    )


if __name__ == "__main__":
    main()

File: lib/crossword-generator.ts
import { readFile, writeFile } from "node:fs/promises";
import path from "node:path";

import wordDictionary from "../scripts/slownik_bez_rzadkich.json";

const SIZE = 5;
const MIN_WORD_LENGTH = 3;
const MAX_ATTEMPTS = 500;

type Direction = "horizontal" | "vertical";

type Slot = {
    row: number;
    col: number;
    length: number;
    direction: Direction;
};

type ClueCache = Record<string, string | null>;

export type GeneratedCrossword = {
    size: number;
    grid: string[][];
    words: Array<{
        word: string;
        row: number;
        col: number;
        length: number;
        direction: Direction;
        clue: string;
    }>;
};

class SeededRandom {
    private state: number;

    constructor(seed: number) {
        this.state = seed || 1;
    }

    next(): number {
        this.state = (this.state * 1664525 + 1013904223) >>> 0;
        return this.state / 4294967296;
    }

    integer(maxExclusive: number): number {
        return Math.floor(this.next() * maxExclusive);
    }

    shuffle<T>(values: T[]): void {
        for (let index = values.length - 1; index > 0; index -= 1) {
            const swapIndex = this.integer(index + 1);
            [values[index], values[swapIndex]] = [values[swapIndex], values[index]];
        }
    }
}

function getWords(): string[] {
    const words = Object.values(wordDictionary).flatMap((lengths) => [
        ...lengths["3"],
        ...lengths["4"],
        ...lengths["5"],
    ]);

    return [...new Set(words.map((word) => word.trim().toLowerCase()))].filter(
        (word) =>
            word.length >= MIN_WORD_LENGTH &&
            word.length <= SIZE &&
            /^[a-ząćęłńóśźż]+$/i.test(word),
    );
}

function getCells(slot: Slot): Array<[number, number]> {
    return Array.from({ length: slot.length }, (_, index) =>
        slot.direction === "horizontal"
            ? [slot.row, slot.col + index]
            : [slot.row + index, slot.col],
    );
}

function generateBlockPattern(random: SeededRandom): boolean[][] {
    const blocks = Array.from({ length: SIZE }, () =>
        Array.from({ length: SIZE }, () => false),
    );
    const candidates: Array<[number, number]> = [];

    for (let row = 0; row < SIZE; row += 1) {
        for (let col = 0; col < SIZE; col += 1) {
            const mirroredRow = SIZE - 1 - row;
            const mirroredCol = SIZE - 1 - col;
            if (row === 2 && col === 2) continue;
            if (row > mirroredRow || (row === mirroredRow && col > mirroredCol)) continue;
            candidates.push([row, col]);
        }
    }

    random.shuffle(candidates);
    const blockCount = random.integer(7);
    for (let index = 0; index < blockCount; index += 1) {
        const [row, col] = candidates[index];
        blocks[row][col] = true;
        blocks[SIZE - 1 - row][SIZE - 1 - col] = true;
    }
    return blocks;
}

function extractSlots(blocks: boolean[][]): Slot[] {
    const slots: Slot[] = [];

    for (let row = 0; row < SIZE; row += 1) {
        let col = 0;
        while (col < SIZE) {
            if (blocks[row][col]) {
                col += 1;
                continue;
            }
            const start = col;
            while (col < SIZE && !blocks[row][col]) col += 1;
            if (col - start >= MIN_WORD_LENGTH) {
                slots.push({ row, col: start, length: col - start, direction: "horizontal" });
            }
        }
    }

    for (let col = 0; col < SIZE; col += 1) {
        let row = 0;
        while (row < SIZE) {
            if (blocks[row][col]) {
                row += 1;
                continue;
            }
            const start = row;
            while (row < SIZE && !blocks[row][col]) row += 1;
            if (row - start >= MIN_WORD_LENGTH) {
                slots.push({ row: start, col, length: row - start, direction: "vertical" });
            }
        }
    }

    return slots;
}

function validBlockPattern(blocks: boolean[][]): boolean {
    return extractSlots(blocks).length > 0 && [
        ...Array.from({ length: SIZE }, (_, row) => {
            const lengths: number[] = [];
            let length = 0;
            for (let col = 0; col < SIZE; col += 1) {
                if (blocks[row][col]) {
                    if (length > 0) lengths.push(length);
                    length = 0;
                } else length += 1;
            }
            if (length > 0) lengths.push(length);
            return lengths;
        }),
        ...Array.from({ length: SIZE }, (_, col) => {
            const lengths: number[] = [];
            let length = 0;
            for (let row = 0; row < SIZE; row += 1) {
                if (blocks[row][col]) {
                    if (length > 0) lengths.push(length);
                    length = 0;
                } else length += 1;
            }
            if (length > 0) lengths.push(length);
            return lengths;
        }),
    ].flat().every((length) => length === 0 || length >= MIN_WORD_LENGTH);
}

function getCandidates(slot: Slot, grid: (string | null)[][], words: string[]): string[] {
    return words.filter((word) => {
        if (word.length !== slot.length) return false;
        return getCells(slot).every(([row, col], index) => {
            const existing = grid[row][col];
            return existing === null || existing === word[index];
        });
    });
}

function solve(
    slots: Slot[],
    grid: (string | null)[][],
    words: string[],
    usedWords: Set<string>,
    random: SeededRandom,
): boolean {
    if (slots.length === 0) return true;

    let bestSlots: Slot[] = [];
    let bestCandidates: string[] = [];

    for (const slot of slots) {
        const candidates = getCandidates(slot, grid, words).filter(
            (word) => !usedWords.has(word),
        );
        if (bestSlots.length === 0 || candidates.length < bestCandidates.length) {
            bestSlots = [slot];
            bestCandidates = candidates;
        } else if (candidates.length === bestCandidates.length) {
            bestSlots.push(slot);
        }
        if (candidates.length === 0) return false;
    }

    const slot = bestSlots[random.integer(bestSlots.length)];
    bestCandidates = getCandidates(slot, grid, words).filter(
        (word) => !usedWords.has(word),
    );
    random.shuffle(bestCandidates);

    for (const word of bestCandidates) {
        const changed: Array<[number, number]> = [];
        for (const [index, [row, col]] of getCells(slot).entries()) {
            if (grid[row][col] === null) {
                grid[row][col] = word[index];
                changed.push([row, col]);
            }
        }
        usedWords.add(word);
        if (solve(slots.filter((candidate) => candidate !== slot), grid, words, usedWords, random)) {
            return true;
        }
        usedWords.delete(word);
        for (const [row, col] of changed) grid[row][col] = null;
    }
    return false;
}

const clueCachePath = path.join(process.cwd(), "scripts", "clue_cache.json");

function decodeHtml(value: string): string {
    return value
        .replace(/<[^>]+>/g, " ")
        .replace(/&nbsp;/gi, " ")
        .replace(/&amp;/gi, "&")
        .replace(/&quot;/gi, '"')
        .replace(/&#39;|&apos;/gi, "'")
        .replace(/&lt;/gi, "<")
        .replace(/&gt;/gi, ">")
        .replace(/&#(\d+);/g, (_, code: string) => String.fromCodePoint(Number(code)))
        .replace(/\s+/g, " ")
        .trim();
}

function cleanClue(value: string, word: string): string | null {
    const clue = decodeHtml(value).replace(/^\s*(?:znaczenie|definicja)\s*:\s*/i, "");
    if (clue.length < 5 || clue.toLowerCase() === word || clue.toLowerCase().startsWith(`${word} `)) return null;
    return clue.length > 120 ? `${clue.slice(0, 117).trimEnd()}...` : clue;
}

async function fetchText(url: string): Promise<string | null> {
    try {
        const response = await fetch(url, {
            headers: { "User-Agent": "Polaczenia crossword generator/1.0" },
            signal: AbortSignal.timeout(5000),
        });
        return response.ok ? response.text() : null;
    } catch {
        return null;
    }
}

async function fetchSjpClue(word: string): Promise<string | null> {
    const html = await fetchText(`https://sjp.pl/${encodeURIComponent(word)}`);
    if (!html) return null;
    const markerIndex = html.search(/znaczenie\s*:/i);
    if (markerIndex < 0) return null;
    const section = html.slice(markerIndex, markerIndex + 12000);
    const candidates = section.match(/<(?:p|dd|li|div)[^>]*>([\s\S]*?)<\/(?:p|dd|li|div)>/gi) ?? [];
    for (const candidate of candidates) {
        const clue = cleanClue(candidate, word);
        if (clue) return clue;
    }
    return null;
}

async function fetchWiktionaryClue(word: string): Promise<string | null> {
    const url = new URL("https://pl.wiktionary.org/w/api.php");
    url.searchParams.set("action", "parse");
    url.searchParams.set("page", word);
    url.searchParams.set("prop", "text");
    url.searchParams.set("format", "json");
    url.searchParams.set("origin", "*");
    const response = await fetchText(url.toString());
    if (!response) return null;
    try {
        const html = (JSON.parse(response) as { parse?: { text?: { "*"?: string } } }).parse?.text?.["*"];
        if (!html) return null;
        const meanings = html.match(/<h[23][^>]*>[\s\S]*?znaczenia[\s\S]*?<\/h[23]>([\s\S]*?)(?=<h[23]|$)/i)?.[1];
        const candidates = (meanings ?? html).match(/<li[^>]*>([\s\S]*?)<\/li>/gi) ?? [];
        for (const candidate of candidates) {
            const clue = cleanClue(candidate, word);
            if (clue) return clue;
        }
    } catch {
        return null;
    }
    return null;
}

async function loadClueCache(): Promise<ClueCache> {
    try {
        return JSON.parse(await readFile(clueCachePath, "utf8")) as ClueCache;
    } catch {
        return {};
    }
}

async function getClue(word: string, cache: ClueCache): Promise<string> {
    if (typeof cache[word] === "string" && cache[word].trim()) return cache[word];

    const clue = await fetchSjpClue(word) ?? await fetchWiktionaryClue(word);
    cache[word] = clue;
    return clue ?? "Brak definicji";
}

async function saveClueCache(cache: ClueCache): Promise<void> {
    try {
        await writeFile(clueCachePath, `${JSON.stringify(cache, null, 2)}\n`, "utf8");
    } catch {
        // Cache jest optymalizacją; brak możliwości zapisu nie powinien blokować generowania.
    }
}

async function createOutput(blocks: boolean[][], slots: Slot[], grid: (string | null)[][]): Promise<GeneratedCrossword> {
    const cache = await loadClueCache();
    const words = slots.map((slot) => getCells(slot).map(([row, col]) => grid[row][col]).join(""));
    const clues = await Promise.all(words.map((word) => getClue(word.toLowerCase(), cache)));
    await saveClueCache(cache);

    return {
        size: SIZE,
        grid: blocks.map((row, rowIndex) =>
            row.map((isBlock, colIndex) => (isBlock ? "#" : grid[rowIndex][colIndex] ?? "#")),
        ),
        words: slots.map((slot, index) => {
            const word = words[index];
            return {
                word,
                row: slot.row,
                col: slot.col,
                length: slot.length,
                direction: slot.direction,
                clue: clues[index],
            };
        }),
    };
}

export async function generateCrossword(seed: number): Promise<GeneratedCrossword> {
    const random = new SeededRandom(seed);
    const words = getWords();

    for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt += 1) {
        const blocks = generateBlockPattern(random);
        if (!validBlockPattern(blocks)) continue;
        const slots = extractSlots(blocks);
        if (!slots.some((slot) => slot.direction === "horizontal") || !slots.some((slot) => slot.direction === "vertical")) continue;

        const grid = blocks.map((row) => row.map((isBlock) => (isBlock ? "#" : null)));
        const usedWords = new Set<string>();
        if (solve(slots, grid, words, usedWords, random)) {
            return createOutput(blocks, slots, grid);
        }
    }

    throw new Error("Nie udało się wygenerować krzyżówki.");
}
